src/Control/Effect/Telemetry.hs

```diff
@@ -0,0 +1,50 @@
+module Control.Effect.Telemetry (
+  Telemetry (..),
+  trackUsage,
+  trackTimeSpent,
+  trackConfig,
+  trackRawLogMessage,
+  setSink,
+  trackResult,
+) where
+
+import App.Fossa.Telemetry.Sink.Common (TelemetrySink)
+import App.Fossa.Telemetry.Types (CountableCliFeature)
+import Control.Algebra (Has, send)
+import Data.Aeson (ToJSON (toJSON), Value)
+import Data.Text (Text)
+import Diag.Result (Result (Failure, Success), renderFailure, renderSuccess)
+import Effect.Logger (Severity (SevError, SevWarn), renderIt, renderItMd)
+import Prettyprinter (Doc)
+import Prettyprinter.Render.Terminal (AnsiStyle)
+
+data Telemetry m a where
+  TrackUsage :: CountableCliFeature -> Telemetry m ()
+  TrackTimeSpent :: Text -> m a -> Telemetry m a
+  TrackConfig :: Text -> Value -> Telemetry m ()
+  TrackRawLogMessage :: Severity -> Text -> Telemetry m ()
+  SetTelemetrySink :: TelemetrySink -> Telemetry m ()
+
+trackUsage :: Has Telemetry sig m => CountableCliFeature -> m ()
+trackUsage = send . TrackUsage
+
+trackTimeSpent :: Has Telemetry sig m => Text -> m a -> m a
+trackTimeSpent header act = send $ TrackTimeSpent header act
+
+trackConfig :: (ToJSON cfg, Has Telemetry sig m) => Text -> cfg -> m ()
+trackConfig cmd config = send . TrackConfig cmd $ toJSON config
+
+{-# WARNING trackRawLogMessage "Only use this if absolutely necessary, you likely want to use 'trackResult'" #-}
```

Comment:
TIL you can leave custom warnings.  However, this presents a problem, because we convert all warnings in the codebase to errors, using `-Werror`.  This function is likely completely unusable as-is, unless those warnings are somehow left out of the `-Werror` conversion.